include/seqan3/io/sam_file/format_sam_base.hpp

```diff
@@ -459,16 +459,34 @@ inline void format_sam_base::read_header(stream_view_type && stream_view,
                                          ref_seqs_type & /*ref_id_to_pos_map*/)
 {
     auto it = std::ranges::begin(stream_view);
+    auto end = std::ranges::end(stream_view);
+    std::string string_buffer{};
 
-    auto parse_tag_value = [&stream_view, &it, this] (auto & value) // helper function to parse the next tag value
+    auto call_read_field = [&it, &string_buffer, this] (auto predicate, auto & value)
     {
-        detail::consume(stream_view | views::take_until_or_throw(is_char<':'>)); // skip tag name
-        std::ranges::next(std::ranges::begin(stream_view));                     // skip ':'
-        read_field(stream_view | views::take_until_or_throw(is_char<'\t'> || is_char<'\n'>), value);
-        it = std::ranges::begin(stream_view);                                    // make sure iterator is up2date
+        string_buffer.clear();
+        while (!predicate(*it))
+        {
+            string_buffer.push_back(*it);
+            ++it;
+        }
+        read_field(string_buffer, value);
+    };
+
+    auto consume = [&it] ()
+    {
+        while (!is_char<':'>(*it))
+            ++it;
+        ++it;
+    };
+
+    auto parse_tag_value = [&] (auto & value) // helper function to parse the next tag value
+    {
+        consume();
+        call_read_field(is_char<'\t'> || is_char<'\n'>, value);
     };
 
-    while (is_char<'@'>(*it))
+    while (it != end && is_char<'@'>(*it))
```

Comment:
```suggestion
    while (it != end && is_char<'@'>(*it))
```

`it != end` is used in BAM, the `stream_view` is a `view_take_exactly_or_throw` and we need to stop at the end (the size is known).

`is_char<'@'>(*it)` is used in SAM, the `stream_view` is a view over the file and we stop when we read all lines that start with `@`.